### PR TITLE
#160 Key Values custom columns number

### PR DESCRIPTION
**Pull request by NFSS10**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-robin-revamp/issues/160 |
| Dependencies | -- |
| Decisions | • Added the optinal boolean prop `twoColumns` to make `KeyValues` have two columns. In tablet it defaults too true <br>• Supports all of the `KeyValue` props |
| Animated GIF | **Two Columns in tablet:**<br>RN-Key_values_only_two_columns<br>**The addition of all the KeyValue props, allowed to have pressable columns as an example:**<br>RN-Key_values_support_for_more_props |

### CODE REVIEW

File: react/components/molecules/key-values/key-values.js [hunk 1]
```diff
@@ -34,4 +78,15 @@ export class KeyValues extends PureComponent {
...
 }
 
+const styles = StyleSheet.create({
+    keyValuesColumns: {
+        flexDirection: "row",
+        flexWrap: "wrap",
+        justifyContent: "space-between"
+    },
+    keyValueTwoColumns: {
+        width: "48%"
```
Comment (gcandal) [inline react/components/molecules/key-values/key-values.js:88]: Gotcha 😅. Why not 50%?

Comment (NFSS10) [reply react/components/molecules/key-values/key-values.js:88]: To have some space between them. I use this in combination with ` justifyContent: "space-between"`.

Margin doesnt count for the component width, so the flexWrap doesnt work properly.
When using padding, the border doesn't shift, stays the same

This was the solution that seemed ok.

Comment (gcandal) [reply react/components/molecules/key-values/key-values.js:88]: Ok I see. As a last check, can you please confirm how is it currently working on the order activity list? There we use `FlatList` but maybe you can take inspiration from there, just to confirm if there's an alternative.

**Comment by NFSS10** [comment]
As discussed, updated with a `View` wrapping the `KeyValue` to be able to had `margin`

**Comment by NFSS10** [comment]
> > As discussed, updated with a `View` wrapping the `KeyValue` to be able to had `margin`
> 
> Does this work correctly? Shouldn't just the items for the right columns (odd indexes) have a margin-left?

Yes
Look at the condition:
`return this.props.twoColumns && index % 2 > 0 ? styles.keyValueTwoColumns : {};`

It has a short circuit condition, has to have the `twoColumns` flag = true to not process the % (module) operation.

and then the % operation itself, which is true only when the index is not one of the first columns
for example:
2 columns, 5 items
0%2 = 0 -> returns {}
1%2 = 1 -> returns style
2%2 = 0 -> returns {}
3%2 = 1 -> returns style
4%2 = 0 -> returns {}

Basically it applies to every column that isn't the first one, we could change `2` to any number of columns and the logic would still work

File: react/components/molecules/key-values/key-values.js [hunk 2]
```diff
@@ -1,37 +1,101 @@
 import React, { PureComponent } from "react";
-import { View, ViewPropTypes } from "react-native";
+import { StyleSheet, View, ViewPropTypes } from "react-native";
 import PropTypes from "prop-types";
 
 import { KeyValue } from "../key-value";
+import { isTabletSize } from "../../../util/ui-utils";
```
Comment (joamag) [inline react/components/molecules/key-values/key-values.js:6]: just `../../../util` right?

Comment (NFSS10) [reply react/components/molecules/key-values/key-values.js]: Yes, fixed the import

File: react/components/molecules/key-values/key-values.js [hunk 3]
```diff
@@ -1,37 +1,101 @@
...
+    _style = () => {
+        return [this.props.twoColumns ? styles.keyValuesColumns : {}, this.props.style];
+    };
+
+    _keyValueWrapperStyle = () => {
+        return this.props.twoColumns ? styles.twoColumns : {};
+    };
+
+    _keyValueStyle = index => {
+        return this.props.twoColumns && index % 2 > 0 ? styles.keyValueTwoColumns : {};
```
Comment (gcandal) [inline react/components/molecules/key-values/key-values.js:51]: `keyValueTwoColumns` seems misleading because the name indicates that it should be applied to all key value when there are two columns, but the truth is it's only applied to the right column. Please rename accordingly.

File: react/components/molecules/key-values/key-values.js [hunk 4]
```diff
@@ -1,37 +1,101 @@
...
+            style: {}
+        };
     }
 
+    _style = () => {
+        return [this.props.twoColumns ? styles.keyValuesColumns : {}, this.props.style];
+    };
+
+    _keyValueWrapperStyle = () => {
+        return this.props.twoColumns ? styles.twoColumns : {};
```
Comment (gcandal) [inline react/components/molecules/key-values/key-values.js:47]: `twoColumns` -> `keyValueTwoColumns`

File: react/components/molecules/key-values/key-values.js (hunk 3 above)
Comment (gcandal) [reply react/components/molecules/key-values/key-values.js:51]: Suggestion: `keyValueColumnRight`
